crates/slab/src/lib.rs

```diff
@@ -0,0 +1,450 @@
+//!     }
+//!
+//!     pub fn get(&self, id: GenerationalId) -> Option<&T> {
+//!         let entry = self.slab.get(id.id)?;
+//!
+//!         // Check that the entry's generation matches the id's generation,
+//!         // else we have an ABA bug. (Alternatively, return `None` instead
+//!         // of panicking.)
+//!         assert_eq!(id.generation, entry.generation);
+//!
+//!         Some(&entry.value)
+//!     }
+//!
+//!     pub fn dealloc(&mut self, id: GenerationalId) {
+//!         // Check that the entry's generation matches the id's generation,
+//!         // else we have an ABA bug. (Alternatively, silently return on
+//!         // double-free instead of panicking.)
+//!         assert_eq!(id.generation, self.slab[id.id].generation);
+//!
+//!         self.slab.dealloc(id.id);
+//!
+//!         // Increment our generation whenever we deallocate so that any new
+//!         // value placed in this same entry will have a different generation
+//!         // and we can detect ABA bugs.
+//!         self.generation += 1;
+//!     }
+//! }
+//! ```
+
+#![forbid(unsafe_code)]
+#![deny(missing_docs, missing_debug_implementations)]
+
+use std::num::NonZeroU32;
+
+/// An identifier for an allocated value inside a `slab`.
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+pub struct Id(EntryIndex);
+
+impl std::fmt::Debug for Id {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("Id").field(&self.0.index()).finish()
+    }
+}
+
+impl Id {
+    /// Get the raw underlying representation of this `Id`.
+    #[inline]
+    pub fn into_raw(self) -> u32 {
+        u32::try_from(self.0.index()).unwrap()
+    }
+
+    /// Construct an `Id` from its raw underlying representation.
+    ///
+    /// `raw` should be a value that was previously created via
+    /// `Id::into_raw`. May panic if given arbitrary values.
+    #[inline]
+    pub fn from_raw(raw: u32) -> Self {
+        let raw = usize::try_from(raw).unwrap();
+        Self(EntryIndex::new(raw))
+    }
+}
+
+/// A simple, uni-typed slab arena.
+pub struct Slab<T> {
+    /// The slab's entries, each is either occupied and holding a `T` or vacant
+    /// and is a link the free list.
+    entries: Vec<Entry<T>>,
+
+    /// The index of the first free entry in the free list.
+    free: Option<EntryIndex>,
+
+    /// The number of occupied entries is this slab.
+    len: u32,
+}
+
+impl<T> std::fmt::Debug for Slab<T>
+where
+    T: std::fmt::Debug,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_map().entries(self.iter()).finish()
+    }
+}
+
+enum Entry<T> {
+    /// An occupied entry holding a `T`.
+    Occupied(T),
+
+    /// A vacant entry.
+    Free {
+        /// A link in the slab's free list, pointing to the next free entry, if
+        /// any.
+        next_free: Option<EntryIndex>,
+    },
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+struct EntryIndex(NonZeroU32);
```

Comment:
Does this use `NonZeroU32` to ensure that `Option<Id>` takes up four bytes?

Reply:
Yep, and this also aligns with how `VMSharedTypeIndex` reserves `u32::MAX` for its default/null value.